src/pybind/mgr/cephadm/inventory.py

```diff
@@ -697,9 +715,65 @@ def save_host(self, host: str) -> None:
             j['scheduled_daemon_actions'] = self.scheduled_daemon_actions[host]
         if host in self.metadata_up_to_date:
             j['metadata_up_to_date'] = self.metadata_up_to_date[host]
+        if host in self.devices:
+            self.save_host_devices(host)
 
         self.mgr.set_store(HOST_CACHE_PREFIX + host, json.dumps(j))
 
+    def save_host_devices(self, host: str) -> None:
+        if host not in self.devices or not self.devices[host]:
+            logger.debug(f'Host {host} has no devices to save')
+            return
+
+        devs: List[Dict[str, Any]] = []
+        for d in self.devices[host]:
+            devs.append(d.to_json())
+
+        def byte_len(s: str) -> int:
+            return len(s.encode('utf-8'))
+
+        dev_cache_counter: int = 0
+        cache_size: int = self.mgr.get_foreign_ceph_option('mon', 'mon_config_key_max_entry_size')
+        if cache_size is not None and cache_size != 0 and byte_len(json.dumps(devs)) > cache_size:
+            # no guarantee all device entries take up the same amount of space
+            # splitting it up so there's one more entry than we need should be fairly
+            # safe and save a lot of extra logic checking sizes
+            cache_entries_needed = math.ceil(byte_len(json.dumps(devs)) / cache_size) + 1
+            dev_sublist_size = math.ceil(len(devs) / cache_entries_needed)
+            dev_lists: List[List[Dict[str, Any]]] = [devs[i:i + dev_sublist_size]
+                                                     for i in range(0, len(devs), dev_sublist_size)]
+            for dev_list in dev_lists:
+                self.mgr.set_store(HOST_CACHE_PREFIX + host + '.devices.'
```

Comment:
Here I have a suggestion which I'm not sure if it make sense or not. As I understand, we are calculating how many cache entries we need to store the devs. Probably it makes sense in this case to store this info in some entry such as: `HOST_CACHE_PREFIX + host + '.devices.info'` this way when reading back the data you first we try to read this entry and if found, then we need for sure many entries we need to read from the cache. In addition, doing it this way would make the `load_host_devices` more explicit (since instead of using a while it would use a finite loop. Another advantage is you would create explicit UT also for this code.